Documentation/Evolution/RegexSyntax.md

```diff
@@ -0,0 +1,848 @@
+
+```
+Anchor -> '^' | '$' | '\A' | '\b' | '\B' | '\G' | '\y' | '\Y' | '\z' | '\Z'
+```
+
+Anchors match against a certain position in the input rather than on a particular character of the input.
+
+- `^`: Matches at the start of a line.
+- `$`: Matches at the end of a line.
+- `\A`: Matches at the very start of the input string.
+- `\Z`: Matches at the very end of the input string, in addition to before a newline at the very end of the input string.
+- `\z`: Like `\Z`, but only matches at the very end of the input string.
+- `\G`: Like `\A`, but also matches against the start position of where matching resumes in global matching mode (e.g `\Gab` matches twice in `abab`, `\Aab` would only match once).
+- `\b` matches a boundary between a word character and a non-word character. The definitions of which vary depending on matching engine.
+- `\B` matches a non-word-boundary.
+- `\y` matches a text segment boundary, the definition of which varies based on the `y{w}` and `y{g}` matching option.
+- `\Y` matches a non-text-segment-boundary.
+
+#### Escape sequences
+
+```
+EscapeSequence -> '\a' | '\b' | '\c' <Char> | '\e' | '\f' | '\n' | '\r' | '\t'
+```
+
+These escape sequences each denote a specific scalar value.
+
+- `\a`: The alert (bell) character `U+7`.
+- `\b`: The backspace character `U+8`. Note this may only be used in a custom character class, otherwise it represents a word boundary.
+- `\c <Char>`: A control character sequence, which denotes a scalar from `U+00` - `U+7F` depending on the ASCII character provided.
+- `\e`: The escape character `U+1B`.
+- `\f`: The form-feed character `U+C`.
+- `\n`: The newline character `U+A`.
+- `\r`: The carriage return character `U+D`.
+- `\t`: The tab character `U+9`.
+
+#### Builtin character classes
+
+```
+BuiltinCharClass -> '.' | '\C' | '\d' | '\D' | '\h' | '\H' | '\N' | '\O' | '\R' | '\s' | '\S' | '\v' | '\V' | '\w' | '\W' | '\X'
+```
+
+- `.`: Any character excluding newlines.
+- `\C`: A single UTF code unit.
+- `\d`: Digit character.
+- `\D`: Non-digit character.
+- `\h`: Horizontal space character.
+- `\H`: Non-horizontal-space character.
+- `\N`: Non-newline character.
+- `\O`: Any character (including newlines). This is syntax from Oniguruma.
+- `\R`: Newline sequence.
+- `\s`: Whitespace character.
+- `\S`: Non-whitespace character.
+- `\v`: Vertical space character.
+- `\V`: Non-vertical-space character.
+- `\w`: Word character.
+- `\W`: Non-word character.
+- `\X`: Any extended grapheme cluster.
+
+Precise definitions of character classes is discussed in [Character Classes for String Processing](https://forums.swift.org/t/pitch-character-classes-for-string-processing/52920).
+
+#### Unicode scalars
+
+```
+UnicodeScalar -> '\u{' HexDigit{1...} '}'
+               | '\u'  HexDigit{4}
+               | '\x{' HexDigit{1...} '}'
+               | '\x'  HexDigit{0...2}
+               | '\U'  HexDigit{8}
+               | '\o{' OctalDigit{1...} '}'
+               | '\0' OctalDigit{0...3}
```

Comment:
In earlier discussions, we had talked about only supporting the `'\u{' HexDigit{1...} '}'` style for unicode scalar literals, to match strings, and offering fixits for all the others. Are we planning to support all these varied styles instead?

Reply:
I believe we want to support at least parsing them, but we will likely want to warn and tell users to use `\u{...}` instead

Reply:
"Support" should be maximally permissive (balanced with being clear and  unambiguous) as that's the main value of going with existing syntax instead of a novel syntax. But, we can also negotiate what is warned on and what has fixits.

I think the "Swift canonical" scalar syntax is clearly superior and should be encouraged, but there's a choice of mechanism here. We certainly should accept the other forms and compile them to the same thing in the end, it's a question of whether and what kind of diagnostics we emit.

I am a little concerned with tons of warnings, especially since Swift doesn't support fine-grained warning control (err, it does internally because I added it way back when when I added global warning suppression and warnings-as-errors, but I don't think it's been exposed). A bad developer experience is warnings that are unavoidable or for whom addressing is worse than living with them.

For a function taking a run-time string containing regex syntax, we'd support all the ways to spell it rather than throw a cannot-compile error. A different API would be whether it's in Swift canonical form or to put it into Swift canonical form. If we warn too heavily, that could even encourage people to use raw string literals instead of regex literals, which would be awkward.

Reply:
FWIW, I disagree about accepting the other forms. This list is a superset; AFAIK there's no language that supports the full laundry list. That means that whenever you're moving regexes from language to language there's some necessary massaging of things like this. If we can make that as easy as clicking a fix-it or pressing ⌃⌥⌘F, that's more support than you typically get.